src/connection-state-handler.ts

```diff
@@ -0,0 +1,116 @@
+import { EventEmitter, EventMap } from './event-emitter';
+import { logger } from './util/logger';
+
+// Overall connection state (based on the ICE and DTLS connection states)
+export enum ConnectionState {
+  New = 'New', // connection attempt has not been started
+  Closed = 'Closed', // connection closed, there is no way to move out of this state
+  Connected = 'Connected', // both ICE and DTLS connections are established, media is flowing
+  Connecting = 'Connecting', // initial connection attempt in progress
+  Disconnected = 'Disconnected', // connection lost temporarily, the browser is trying to re-establish it automatically
+  Failed = 'Failed', // connection failed, an ICE restart is required
+}
+
+enum ConnectionStateEvents {
+  ConnectionStateChanged = 'ConnectionStateChanged',
+}
+
+interface ConnectionStateEventHandlers extends EventMap {
+  [ConnectionStateEvents.ConnectionStateChanged]: (state: ConnectionState) => void;
+}
+
+type GetCurrentStatesCallback = () => {
+  connectionState: RTCPeerConnectionState;
+  iceState: RTCIceConnectionState;
+};
+
+/**
+ * Listens on the connection's ICE and DTLS state changes and emits a single
+ * event that summarizes all the internal states into a single overall connection state.
+ */
+export class ConnectionStateHandler extends EventEmitter<ConnectionStateEventHandlers> {
+  static Events = ConnectionStateEvents;
+
+  private mediaConnectionState: ConnectionState;
+
+  private getCurrentStatesCallback: GetCurrentStatesCallback;
+
+  /**
+   * Creates an instance of ConnectionStateHandler.
+   *
+   * @param getCurrentStatesCallback - Callback for getting the connection state information
+   *                                   from the peer connection.
+   */
+  constructor(getCurrentStatesCallback: GetCurrentStatesCallback) {
+    super();
+    this.getCurrentStatesCallback = getCurrentStatesCallback;
+    this.mediaConnectionState = this.evaluateMediaConnectionState();
+  }
+
+  /**
+   * Handler for connection state change.
+   */
+  public onConnectionStateChange(): void {
+    const newConnectionState = this.evaluateMediaConnectionState();
+
+    if (newConnectionState !== this.mediaConnectionState) {
+      this.mediaConnectionState = newConnectionState;
+      this.emit(ConnectionStateEvents.ConnectionStateChanged, this.mediaConnectionState);
+    }
+  }
+
+  /**
+   * Handler for ice connection state change.
+   */
+  public onIceConnectionStateChange(): void {
+    const newConnectionState = this.evaluateMediaConnectionState();
+
+    if (newConnectionState !== this.mediaConnectionState) {
+      this.mediaConnectionState = newConnectionState;
+      this.emit(ConnectionStateEvents.ConnectionStateChanged, this.mediaConnectionState);
+    }
+  }
```

Comment:
It looks like this and `onConnectionStateChange` could both leverage some private helper method?

Reply:
done